File: projects/mvcc/internal/transaction/transaction.go
package transaction

import (
	"fmt"
	"sync"
	"sync/atomic"
)

// TxnStatus represents the status of a transaction
type TxnStatus int

const (
	TxnActive    TxnStatus = iota // Transaction is active
	TxnCommitted                  // Transaction has committed
	TxnAborted                    // Transaction has been aborted
)

func (s TxnStatus) String() string {
	switch s {
	case TxnActive:
		return "Active"
	case TxnCommitted:
		return "Committed"
	case TxnAborted:
		return "Aborted"
	default:
		return "Unknown"
	}
}

// WriteRecord tracks a write operation within a transaction
type WriteRecord struct {
	Key      string
	Value    []byte
	IsDelete bool
}

// Transaction represents a single MVCC transaction
type Transaction struct {
	ID              uint64
	StartTimestamp  uint64    // Snapshot timestamp when transaction began
	CommitTimestamp uint64    // Timestamp assigned at commit time (0 until committed)
	Status          TxnStatus
	ReadSet         map[string]uint64    // key -> read timestamp
	WriteSet        map[string]WriteRecord // key -> write record
	mu              sync.RWMutex
}

// NewTransaction creates a new transaction with the given ID and start timestamp
func NewTransaction(id uint64, startTimestamp uint64) *Transaction {
	return &Transaction{
		ID:             id,
		StartTimestamp: startTimestamp,
		Status:         TxnActive,
		ReadSet:        make(map[string]uint64),
		WriteSet:       make(map[string]WriteRecord),
	}
}

// AddRead records a read operation on the given key
func (t *Transaction) AddRead(key string, readTimestamp uint64) {
	t.mu.Lock()
	defer t.mu.Unlock()
	t.ReadSet[key] = readTimestamp
}

// AddWrite records a write operation on the given key
func (t *Transaction) AddWrite(key string, value []byte) {
	t.mu.Lock()
	defer t.mu.Unlock()
	t.WriteSet[key] = WriteRecord{
		Key:   key,
		Value: value,
	}
}

// AddDelete records a delete operation on the given key
func (t *Transaction) AddDelete(key string) {
	t.mu.Lock()
	defer t.mu.Unlock()
	t.WriteSet[key] = WriteRecord{
		Key:      key,
		IsDelete: true,
	}
}

// GetWrite returns the write record for a key if it exists in the write set
func (t *Transaction) GetWrite(key string) (WriteRecord, bool) {
	t.mu.RLock()
	defer t.mu.RUnlock()
	wr, ok := t.WriteSet[key]
	return wr, ok
}

// IsActive returns true if the transaction is still active
func (t *Transaction) IsActive() bool {
	t.mu.RLock()
	defer t.mu.RUnlock()
	return t.Status == TxnActive
}

// String returns a string representation of the transaction
func (t *Transaction) String() string {
	t.mu.RLock()
	defer t.mu.RUnlock()
	return fmt.Sprintf("Txn{ID: %d, StartTS: %d, CommitTS: %d, Status: %s, Reads: %d, Writes: %d}",
		t.ID, t.StartTimestamp, t.CommitTimestamp, t.Status, len(t.ReadSet), len(t.WriteSet))
}

// TransactionManager manages all transactions and provides timestamp allocation
type TransactionManager struct {
	mu               sync.RWMutex
	clock            uint64 // Logical clock for timestamp allocation
	activeTxns       map[uint64]*Transaction // txnID -> Transaction
	committedTxns    map[uint64]*Transaction // txnID -> Transaction (for GC visibility)
	minActiveTS      uint64 // Minimum start timestamp among active transactions
	needsMinRecalc   bool   // Flag to recalculate minActiveTS
}

// NewTransactionManager creates a new transaction manager
func NewTransactionManager() *TransactionManager {
	return &TransactionManager{
		activeTxns:    make(map[uint64]*Transaction),
		committedTxns: make(map[uint64]*Transaction),
	}
}

// AllocateTimestamp atomically increments and returns a new timestamp
func (tm *TransactionManager) AllocateTimestamp() uint64 {
	return atomic.AddUint64(&tm.clock, 1)
}

// CurrentTimestamp returns the current clock value without incrementing
func (tm *TransactionManager) CurrentTimestamp() uint64 {
	return atomic.LoadUint64(&tm.clock)
}

// Begin starts a new transaction and returns it
func (tm *TransactionManager) Begin() *Transaction {
	ts := tm.AllocateTimestamp()
	txn := NewTransaction(ts, ts)

	tm.mu.Lock()
	tm.activeTxns[txn.ID] = txn
	tm.updateMinActiveTS()
	tm.mu.Unlock()

	return txn
}

// Commit commits a transaction with a new commit timestamp
// Returns error if the transaction is not active
func (tm *TransactionManager) Commit(txnID uint64) (uint64, error) {
	tm.mu.Lock()
	defer tm.mu.Unlock()

	txn, ok := tm.activeTxns[txnID]
	if !ok {
		return 0, fmt.Errorf("transaction %d not found or not active", txnID)
	}

	commitTS := tm.AllocateTimestamp()
	txn.CommitTimestamp = commitTS
	txn.Status = TxnCommitted

	// Move from active to committed
	delete(tm.activeTxns, txnID)
	tm.committedTxns[txnID] = txn
	tm.updateMinActiveTS()

	return commitTS, nil
}

// Abort aborts a transaction
func (tm *TransactionManager) Abort(txnID uint64) error {
	tm.mu.Lock()
	defer tm.mu.Unlock()

	txn, ok := tm.activeTxns[txnID]
	if !ok {
		return fmt.Errorf("transaction %d not found or not active", txnID)
	}

	txn.Status = TxnAborted
	delete(tm.activeTxns, txnID)
	tm.updateMinActiveTS()

	return nil
}

// GetTransaction returns a transaction by ID (checks both active and committed)
func (tm *TransactionManager) GetTransaction(txnID uint64) (*Transaction, bool) {
	tm.mu.RLock()
	defer tm.mu.RUnlock()

	if txn, ok := tm.activeTxns[txnID]; ok {
		return txn, true
	}
	if txn, ok := tm.committedTxns[txnID]; ok {
		return txn, true
	}
	return nil, false
}

// IsActive checks if a transaction is still active
func (tm *TransactionManager) IsActive(txnID uint64) bool {
	tm.mu.RLock()
	defer tm.mu.RUnlock()
	_, ok := tm.activeTxns[txnID]
	return ok
}

// ActiveTransactions returns a list of all active transaction IDs
func (tm *TransactionManager) ActiveTransactions() []uint64 {
	tm.mu.RLock()
	defer tm.mu.RUnlock()

	ids := make([]uint64, 0, len(tm.activeTxns))
	for id := range tm.activeTxns {
		ids = append(ids, id)
	}
	return ids
}

// MinActiveTimestamp returns the minimum start timestamp among active transactions
// This is used by the garbage collector to determine safe cleanup points
func (tm *TransactionManager) MinActiveTimestamp() uint64 {
	tm.mu.RLock()
	defer tm.mu.RUnlock()
	return tm.minActiveTS
}

// ActiveCount returns the number of active transactions
func (tm *TransactionManager) ActiveCount() int {
	tm.mu.RLock()
	defer tm.mu.RUnlock()
	return len(tm.activeTxns)
}

// updateMinActiveTS recalculates the minimum active timestamp
// Must be called with tm.mu held
func (tm *TransactionManager) updateMinActiveTS() {
	if len(tm.activeTxns) == 0 {
		tm.minActiveTS = tm.CurrentTimestamp()
		return
	}

	minTS := uint64(^uint64(0)) // Max uint64
	for _, txn := range tm.activeTxns {
		if txn.StartTimestamp < minTS {
			minTS = txn.StartTimestamp
		}
	}
	tm.minActiveTS = minTS
}

File: projects/mvcc/src/mvcc/transaction.py
"""
事务管理器 - MVCC 事务的生命周期管理。

事务状态机:
    ACTIVE  ──commit()──>  COMMITTED
    ACTIVE  ──abort()───>  ABORTED

每个事务维护:
    - 读集合 (read_set): 事务读取过的 key 集合
    - 写缓冲 (write_buffer): 事务写入的 key -> value 映射
    - 删除集合 (delete_set): 事务删除的 key 集合
    - 快照 (snapshot): 事务开始时的数据库快照

设计要点:
    1. 事务 ID 全局唯一，由 TransactionManager 统一分配
    2. 事务采用乐观并发控制：写入先缓存，提交时才检测冲突
    3. 支持事务隔离级别: Snapshot Isolation (SI)
"""

from __future__ import annotations
from enum import Enum, auto
from typing import Any, Dict, Optional, Set
from .snapshot import Snapshot


class TxnStatus(Enum):
    """事务状态枚举。"""
    ACTIVE = auto()      # 活跃中
    COMMITTED = auto()   # 已提交
    ABORTED = auto()     # 已中止


class Transaction:
    """
    事务对象 - 代表一个 MVCC 事务。

    属性:
        txn_id:         事务唯一 ID
        start_ts:       事务开始时间戳
        status:         事务状态
        snapshot:       事务快照（开始时的数据库状态）
        read_set:       读集合（用于冲突检测）
        write_buffer:   写缓冲（暂存写入，提交时批量应用）
        delete_set:     删除集合
    """

    def __init__(self, txn_id: int, start_ts: int, snapshot: Snapshot):
        self.txn_id = txn_id
        self.start_ts = start_ts
        self.status = TxnStatus.ACTIVE
        self.snapshot = snapshot
        self.read_set: Set[str] = set()
        self.write_buffer: Dict[str, Any] = {}
        self.delete_set: Set[str] = set()

    def add_read(self, key: str) -> None:
        """记录读操作。"""
        self.read_set.add(key)

    def add_write(self, key: str, value: Any) -> None:
        """记录写操作到写缓冲。"""
        self.write_buffer[key] = value
        # 如果之前标记过删除，取消删除
        self.delete_set.discard(key)

    def add_delete(self, key: str) -> None:
        """记录删除操作。"""
        self.delete_set.add(key)
        # 如果之前有写缓冲，清除
        self.write_buffer.pop(key, None)

    def get_write(self, key: str) -> Optional[Any]:
        """从写缓冲中获取值。"""
        return self.write_buffer.get(key)

    def is_write_buffered(self, key: str) -> bool:
        """检查 key 是否在写缓冲中。"""
        return key in self.write_buffer or key in self.delete_set

    def commit(self, commit_ts: int) -> None:
        """提交事务。"""
        if self.status != TxnStatus.ACTIVE:
            raise RuntimeError(
                f"Cannot commit transaction {self.txn_id}: "
                f"status is {self.status.name}"
            )
        self.status = TxnStatus.COMMITTED
        self._commit_ts = commit_ts

    def abort(self) -> None:
        """中止事务。"""
        if self.status != TxnStatus.ACTIVE:
            raise RuntimeError(
                f"Cannot abort transaction {self.txn_id}: "
                f"status is {self.status.name}"
            )
        self.status = TxnStatus.ABORTED
        # 清理写缓冲
        self.write_buffer.clear()
        self.delete_set.clear()

    @property
    def commit_ts(self) -> Optional[int]:
        """获取提交时间戳（仅已提交事务有值）。"""
        return getattr(self, "_commit_ts", None)

    @property
    def is_active(self) -> bool:
        return self.status == TxnStatus.ACTIVE

    @property
    def is_committed(self) -> bool:
        return self.status == TxnStatus.COMMITTED

    @property
    def is_aborted(self) -> bool:
        return self.status == TxnStatus.ABORTED

    def __repr__(self) -> str:
        return (
            f"Transaction(id={self.txn_id}, "
            f"start_ts={self.start_ts}, "
            f"status={self.status.name}, "
            f"reads={len(self.read_set)}, "
            f"writes={len(self.write_buffer)})"
        )


class TransactionManager:
    """
    事务管理器 - 负责事务的创建、提交、中止和状态查询。

    职责:
        1. 分配全局唯一事务 ID
        2. 维护全局时间戳
        3. 管理活跃事务集合
        4. 查询事务状态
    """

    def __init__(self):
        self._next_txn_id: int = 1
        self._timestamp: int = 0
        self._active_txns: Dict[int, Transaction] = {}
        self._committed_txns: Dict[int, Transaction] = {}
        self._aborted_txns: Dict[int, Transaction] = {}

    def begin(self, snapshot: Snapshot) -> Transaction:
        """
        开始一个新事务。

        Args:
            snapshot: 事务开始时的数据库快照

        Returns:
            新的事务对象
        """
        txn_id = self._next_txn_id
        self._next_txn_id += 1

        start_ts = self.advance_timestamp()
        txn = Transaction(txn_id, start_ts, snapshot)
        self._active_txns[txn_id] = txn
        return txn

    def commit(self, txn: Transaction, commit_ts: int) -> None:
        """提交事务。"""
        txn.commit(commit_ts)
        self._active_txns.pop(txn.txn_id, None)
        self._committed_txns[txn.txn_id] = txn

    def abort(self, txn: Transaction) -> None:
        """中止事务。"""
        txn.abort()
        self._active_txns.pop(txn.txn_id, None)
        self._aborted_txns[txn.txn_id] = txn

    def advance_timestamp(self) -> int:
        """推进全局时间戳并返回新值。"""
        self._timestamp += 1
        return self._timestamp

    def get_timestamp(self) -> int:
        """获取当前时间戳（不推进）。"""
        return self._timestamp

    def get_active_txn_ids(self) -> set:
        """获取所有活跃事务的 ID 集合。"""
        return set(self._active_txns.keys())

    def get_active_txns(self) -> Dict[int, Transaction]:
        """获取所有活跃事务。"""
        return dict(self._active_txns)

    def get_committed_txns(self) -> Dict[int, Transaction]:
        """获取所有已提交事务。"""
        return dict(self._committed_txns)

    def get_committed_txn_ids(self) -> set:
        """获取所有已提交事务的 ID 集合。"""
        return set(self._committed_txns.keys())

    def get_txn(self, txn_id: int) -> Optional[Transaction]:
        """根据 ID 获取事务。"""
        if txn_id in self._active_txns:
            return self._active_txns[txn_id]
        if txn_id in self._committed_txns:
            return self._committed_txns[txn_id]
        if txn_id in self._aborted_txns:
            return self._aborted_txns[txn_id]
        return None

    @property
    def active_count(self) -> int:
        """活跃事务数量。"""
        return len(self._active_txns)

    @property
    def committed_count(self) -> int:
        """已提交事务数量。"""
        return len(self._committed_txns)

    def __repr__(self) -> str:
        return (
            f"TransactionManager(ts={self._timestamp}, "
            f"active={self.active_count}, "
            f"committed={self.committed_count})"
        )
